draft: considerations when building embedded databases (#482)
memfault/interrupt: _drafts/choosing_or_building_an_embedded_db.md

presents some mantainability challenges.

## Aside: Portability of Records

Thinking ahead, you may eventually wish to run your fishtank code on 
different hardware either because of an upgrade to the product,
a new version of the product, or because you want to be able to load
up records from the embedded system inside a mobile app or a server.

That new processor may be 64-bit processor, or may even use a 
different endianness than the microcontroller. Taking a closer look at
the `fishtank_config_t` struct, we see that the basic `int` type is
used, which uses 8 bytes on 64-bit versus 4 bytes on 32-bit,
and the endianness (byte order) depends on the processor.

### Integer Size and Alignment

With the definition of `fishtank_config_t` above,
`sizeof(fishtank_config_t)` will be different for different
platforms. Compiling on a 64-bit MacBook where Apple
kept `int` as 32-bit with 32-bit alignment, you'll get the 
same 8-byte structure as on a 32-bit microcontroller.
However, if you compile on an AWS EC2 instance
(as you may want to do if there is a cloud component to 
your fishtank), then you get a 16-byte structure because
`int` there is 8 bytes.

> The bottom line for portability is, to use the definitions in `stdint.h`
> and instruct the compiler to pack your structure to remove any padding
> which would otherwise be added to encourage alignment to word boundries:

Reviewer: ```suggestion
For portability, use the fixed-width definitions in `stdint.h`
and instruct the compiler to pack your structure to remove any padding
which would otherwise be added to encourage alignment to word boundries:
```

```c
#pragma pack(push, 1)
typedef struct {
    int32_t salinity;
    int32_t oxygen;
} fishtank_config_t;
#pragma pack(pop)
```

### Much Ado About Endianness

> There's one remaining possible portability issue. There have historically 

Reviewer: ```suggestion
There's one remaining portability issue. There have historically 
```

been "big-endian" processors which represent multi-byte integers with the
most-significant byte first rather than least-significant byte first
(so-called "little endian") which has become the dominant endianness today.
This means that even with the fixed-sized, packed structure above,
there may be processors where a record with `salinity == 2` is intepreted

> as `33554432` because of the endian mismatch.

Reviewer: Might be worth a simple example of both to illustrate, though most of the audience should be familiar:

- Decimal value: 1234
- Hexadecimal: 0x04D2 (16-bit integer)
- Big-endian representation:

  |Address|Value|
  |---|---|
  |0x0|0x04 (LSB)|
  |0x1|0xD2 (MSB)|

- Little-endian representation:

  |Address|Value|
  |---|---|
  |0x0|0xD2 (MSB)|
  |0x1|0x04 (LSB)|

Author: Certainly. It's so easy to get it mixed up ;)

DONE (https://github.com/bondhome/interrupt/commit/9cfe15247c544a52f1bc95268e90ba0077175d05)

And although single bits may be programmed, only entire **pages** of flash
may be erased.

3. These pages vary in size between flash chips and microcontrollers, from 512 bytes to 32k, with 4k being a typical size.

4. Flash operations must be **aligned** to specific offsets, usually 32 bytes but this varies.
   The page size, alignment, and possibly restricted read/write lengths together 
   are referred to as the flash **geometry**.

5. Flash has a limited number of program-erase cycles, typically 10k to 100k.
So, to avoid premature failure, 
values that change many times per day must be distributed across different flash pages,
a scheme known as "wear leveling".

6. Flash reading, programming, and especially erasing, is much slower than RAM operations.
So frequently-accessed data must at least be cached in RAM.

These challenges mean that for non-trivial applications such as the
smart home hub where there are many small peices of information to persist,
there is a clear need for a well-organized scheme for managing the use of the 
flash to save and load this data. This is what we refer to as an **embedded database**.

> # Considering your Requirements for an Embedded Database

Reviewer: ```suggestion
## Considering your Requirements for an Embedded Database
```

couple of heading levels need tweaking

Now that it's clear why you need an embedded database, let's consider the different
constraints which may lead you in one direction or another when shopping or building 
a solution:

> #### How much data do you have to store? And how much flash is available?

Reviewer: ```suggestion
### How much data do you have to store? And how much flash is available?
```

etc
